caffe2/onnx/onnx_exporter.cc

```diff
@@ -0,0 +1,585 @@
+ * Copyright (c) 2016-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "caffe2/core/logging.h"
+#include "caffe2/onnx/onnx_exporter.h"
+#include "caffe2/onnx/helper.h"
+#include "caffe2/proto/caffe2_legacy.pb.h"
+#include "caffe2/utils/map_utils.h"
+
+#include <unordered_set>
+
+namespace caffe2 {
+namespace onnx {
+
+namespace {
+void ApplyTrans(
+    std::unordered_map<std::string, AttributeProto>* attrs,
+    bool global,
+    const std::string& k,
+    int dim = 2,
+    const std::string& ks = "") {
+  std::string ks2 = ks.empty() ? (k + "s") : ks;
+  std::string k_h, k_w, k_t, k_l, k_b, k_r;
+  if (dim == 2) {
+    k_h = k + "_h";
+    k_w = k + "_w";
+  } else {
+    k_t = k + "_t";
+    k_l = k + "_l";
+    k_b = k + "_b";
+    k_r = k + "_r";
+  }
+
+  std::vector<int64_t> vals;
+  if (dim == 2 && attrs->count(k_h) && attrs->count(k_w)) {
+    auto it = attrs->find(k_h);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+    it = attrs->find(k_w);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+  } else if (
+      dim == 4 && attrs->count(k_t) && attrs->count(k_b) && attrs->count(k_l) &&
+      attrs->count(k_r)) {
+    auto it = attrs->find(k_t);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+    it = attrs->find(k_l);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+    it = attrs->find(k_b);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+    it = attrs->find(k_r);
+    vals.push_back(it->second.i());
+    attrs->erase(it);
+  } else if (attrs->count(k)) {
+    auto it = attrs->find(k);
+    auto tmp = it->second.i();
+    for (int i = 0; i < dim; ++i) {
+      vals.push_back(tmp);
+    }
+    attrs->erase(it);
+  }
+
+  if (!vals.empty() && !global) {
+    attrs->emplace(ks2, MakeAttribute(ks2, vals));
+  }
+}
+
+int64_t DimProd(const caffe2::TensorShape& shape, int start, int end) {
+  int64_t acc = 1;
+  for (int i = start; i < end; ++i) {
+    acc *= shape.dims(i);
+  }
+  return acc;
+}
+
+TensorProto CreateOnnxShapeTensor(const std::vector<int64_t>& shape) {
+  TensorProto tensor;
+  tensor.set_name(DummyName::NewDummyName());
+  tensor.set_data_type(TensorProto::INT64);
+  tensor.add_dims(shape.size());
+  tensor.mutable_raw_data()->assign(
+      reinterpret_cast<const char*>(shape.data()), sizeof(int64_t) * shape.size());
+  return tensor;
+}
+} // namespace
+
+const std::unordered_map<std::string, std::string>&
```

Comment:
What is the purpose of having functions to get all these maps instead of making them global static objects? Every time you call these functions, they create copies of map + all strings in keys and values. Thats a lot of memory allocations.

Reply:
Why would it create a copy?

Reply:
Right, I didn't notice they return a reference.